### PR TITLE
 [FRONTEND][TFLite] Fully connected op conversion made in sync with TFLite

### PR DESCRIPTION
**Pull request by ANSHUMAN87**
@FrozenGene : Please help review!
Thank you!

### CODE REVIEW

File: tests/python/frontend/tflite/test_forward.py [hunk 1]
```diff
@@ -419,6 +419,31 @@ def test_forward_cast():
...
+
+def test_forward_batch_matmul():
+    """ BATCH_MAT_MUL """
+    print("Jai hanuman!!!")
+    _test_batch_matmul((3, 5, 4), (3, 4, 5), 'float32')
+    _test_batch_matmul((3, 5, 4), (3, 4, 5), 'float32', True, True)
+    _test_batch_matmul((3, 5, 4), (3, 5, 4), 'float32', True, False)
+    _test_batch_matmul((3, 5, 4), (3, 5, 4), 'float32', False, True)
+    _test_batch_matmul((2, 3, 4, 5, 6), (2, 3, 4, 6, 5), 'float32')
+    print("Jai hanuman!!!")
```
Comment (u99127) [inline tests/python/frontend/tflite/test_forward.py:445]: Unneeded print.

File: python/tvm/relay/frontend/tflite.py [hunk 2]
```diff
@@ -1329,16 +1329,20 @@ def convert_fully_connected(self, op):
...
-        assert input_size_per_batch == weight_tensor_shape[1], \
-            "input size and weight size are mismatched"
-        target_shape = tuple((input_tensor_shape[0], input_size_per_batch))
+        # Weight should have only 2 dimensions(TFLite convention)
+        assert len(weight_tensor_shape) == 2, "Weight should be only 2-dim"
+
+        input_size = 1
+        for s in range(0, len(input_tensor_shape)):
+            input_size *= input_tensor_shape[s]
+
```
Comment (u99127) [inline python/tvm/relay/frontend/tflite.py:1338]: A comment here would be appropriate or indeed a reference to TFlite documentation.

File: python/tvm/relay/frontend/tflite.py [hunk 3]
```diff
@@ -1329,16 +1329,20 @@ def convert_fully_connected(self, op):
         input_tensor_shape = input_tensor.tensor.ShapeAsNumpy()
         weight_tensor_shape = weight_tensor.tensor.ShapeAsNumpy()
 
-        # reshape input tensor from N H W C to N H*W*C
-        input_size_per_batch = 1
-        for s in range(1, len(input_tensor_shape)):
-            input_size_per_batch *= input_tensor_shape[s]
-        assert input_size_per_batch == weight_tensor_shape[1], \
-            "input size and weight size are mismatched"
-        target_shape = tuple((input_tensor_shape[0], input_size_per_batch))
```
Comment (u99127) [inline python/tvm/relay/frontend/tflite.py:1338]: Can you explain why this assert is wrong or needs to be removed in your covering note ?

**Comment by ANSHUMAN87** [comment]
> Minor nits, A better commit message would be to refer to handling batch matmul in fully connected op rather than what it stands.
> 
> Isn't there a need for Quantized inputs testing here ?

@u99127 : Thanks for your comments! Just for your info, TFLite converter has multiple limitations in terms of various input types supported, so it is not possible always to add all the cases. 
I had already tried for it, but converter has errors. Also it does not support N-dim.

File: python/tvm/relay/frontend/tflite.py (hunk 3 above)
Comment (ANSHUMAN87) [reply python/tvm/relay/frontend/tflite.py:1338]: @u99127 : This assert is wrong assumption, as it always assumes first dimension is the batch size, while it has utterly failed to link between the input and weights, which is highly dependent when you operate in Fully_Connected or Dense operator.
As the frontend is just a transformation of an successfully deployed model, so we can safely assume the valid relationship exists! 
However this point will be evident when i put some comment to describe batch size computation as per your last comment. Thanks!

**Comment by ANSHUMAN87** [comment]
Gentle ping @FrozenGene !

**Comment by u99127** [comment]
> 
> 
> > Minor nits, A better commit message would be to refer to handling batch matmul in fully connected op rather than what it stands.
> > Isn't there a need for Quantized inputs testing here ?
> 
> @u99127 : Thanks for your comments! Just for your info, TFLite converter has multiple limitations in terms of various input types supported, so it is not possible always to add all the cases.
> I had already tried for it, but converter has errors. 

Thanks for trying it and clarifying the situation with qnn ops.

> Also it does not support N-dim.
> 
> About the commit, actually the changes are quite unrelated to batch matmul, it is just that open up the issue in implementation. So IMO the commit message looks ok to me. Please let me know if you think otherwise. Thanks!


In my experience a good git commit message, explains what was changed and why. It's ok if the body of the commit message is empty as long as the subject line can give a quick and obvious summary to the reader. 

The subject line and the commit message on a quick perusal last night suggested that this commit synchronizes the implementation of fully connected with tflite. Dropping an assert always makes me query why it was dropped which is why I queried it.

regards
Ramana

**Comment by ANSHUMAN87** [comment]
Gentle ping @FrozenGene !

**Comment by FrozenGene** [comment]
Thanks @ANSHUMAN87 @u99127

**Pull request merged**